private/model/api/docstring.go

```diff
@@ -88,21 +92,134 @@ func docstring(doc string) string {
-	doc = wrap(doc, 72)
-
 	return commentify(doc)
 }
 
+// generateDocFromHTML will generate the proper doc string for html encoded doc entries.
+func generateDocFromHTML(htmlSrc string) string {
+	tokenizer := xhtml.NewTokenizer(strings.NewReader(htmlSrc))
+	// tagInfo contains necessary token info of start tag
+	type tagInfo struct {
+		tag string
+		key string
+		val string
+		txt string
+	}
+	doc := ""
+	level := 0
+	col := 0
+	indent := false
+	stack := []tagInfo{}
+
+	for tt := tokenizer.Next(); tt != xhtml.ErrorToken; tt = tokenizer.Next() {
+		switch tt {
+		case xhtml.ErrorToken:
+			break
+		case xhtml.TextToken:
+			txt := string(tokenizer.Text())
+			if len(stack) > 0 {
+				// set the current entries txt value. We use this
+				// for checking if we need to indent. The indent rules
+				// is if there is nested html tags that aren't empty or space,
+				// we will indent the code or li blocks.
+				size := len(stack)
+				stack[size-1].txt = txt
+				info := stack[size-1]
+
+				// if the tag is <code> or <li> we want to indent this block
+				if info.tag == "pre" || info.tag == "li" {
+					indent = true
+				} else if info.tag == "code" {
+					if level == 1 {
+						indent = true
+					} else {
+						for i := level; size-i >= 0 && i > 1; i-- {
+							if stack[size-i].tag == "li" || stack[size-i].tag == "pre" {
+								indent = true
+								break
+							}
+						}
+
+					}
+				}
+
+				// we only want to append to the doc if the current txt isn't
+				// empty or contains only a space.
+				if len(txt) > 0 && txt != " " {
+					if info.tag == "a" {
+						if info.val != "" {
+							txt += fmt.Sprintf(" (%s)", info.val)
+						}
+					}
+
+					// check spacing
+					// we do not care about the current scope we are on, hence the
+					// i > 1. The reason to check for the check against one is due to the
+					// fact that empty stack is at level 0, which we don't care about.
+					for i := level; size-i >= 0 && i > 1; i-- {
+						if stack[size-i].tag == "p" {
+							if len(stack[size-i].txt) > 0 && stack[size-i].txt != " " {
+								indent = false
+							}
+						}
+					}
+
+					txt, col = wrap(txt, col, 72, indent)
+					doc += txt
+				}
+			} else {
+				indent = false
+				txt, col = wrap(txt, col, 72, indent)
+				doc += txt
+			}
+		case xhtml.StartTagToken:
+			tn, _ := tokenizer.TagName()
+			key, val, _ := tokenizer.TagAttr()
+			info := tagInfo{
+				tag: string(tn),
+				key: string(key),
+				val: string(val),
+			}
+			stack = append(stack, info)
+			level++
+		case xhtml.SelfClosingTagToken, xhtml.EndTagToken:
+			// the stack could be empty here
+			indent = false
+			if len(stack) > 0 {
+				info := stack[len(stack)-1]
+				if info.tag == "p" || (len(info.tag) == 2 && info.tag[0] == 'h') {
+					doc += "\n\n"
+					col = 0
+				} else if (level == 1 && info.tag == "code") || info.tag == "pre" || info.tag == "li" {
+					doc += "\n"
+					col = 0
+				}
+				stack = stack[:len(stack)-1]
```

Comment:
ah i think i misread the code, was thinking the `info := stack[len(stack)-1]` and `stack = stack[:len(stack)-1]` were the same subslice, missed that info was an element, not slice.